### ISSUE TITLE
Bug: Hash from a native object that contains an Hash returns unexpected values

### ISSUE THREAD

**Issue by lionelchauvin** [opal/opal#2670]
**Describe the bug**

In ruby the first argument of Hash.new is the default value returned by the hash when a key doesn't exist.
With opal the first argument can be used to initialize the hash from a native javascript object by using backticks:
```
Hash.new(`{a: "A"}`)
```

Opal allows to insert an opal hash in javascript by using #{}:

```
`var h = #{ {a: "A"} }`
```

These two features allow to build an Hash from a native javascipt object with a property that contains an opal hash:
```
h = {a: "A"}
g = Hash.new(`{h: #{h}}`)
```

I am expecting the nested hash to return nil when a key doesn't exist:
```
expect(g['h']['b']).to eq(nil) # because h['b'] == nil
```

but with opal 1.8.2 it returns the nested hash `{"a" => "A"}`.

There is a bug somewhere in
https://github.com/opal/opal/blob/9d48ab627dc253ba26b1507caa768f772c361cbf/stdlib/native.rb#L542

Opal version: 1.8.2

**To Reproduce**

```
# backtick_javascript: true
h = {a: "A"}
g = Hash.new(`{h: #{h}}`)

p g['h']['a'] # "A"
p g['h']['b'] # {"a" => "A"} instead of nil
```

nothing special in generated code:

```
Opal.queue(function(Opal) {/* Generated by Opal 1.8.2 */
  var self = Opal.top, $nesting = [], $$ = Opal.$r($nesting), nil = Opal.nil, h = nil, g = nil;

  Opal.add_stubs('new,p,[]');
  
  h = (new Map([["a", "A"]]));
  g = $$('Hash').$new({h: h});
  self.$p(g['$[]']("h")['$[]']("a"));
  return self.$p(g['$[]']("h")['$[]']("b"));
});

```

**Comment by lionelchauvin** [comment]
I am not sure because it can be called recursively by initialize, but this change seems to fix the problem:

```
diff --git a/stdlib/native.rb b/stdlib/native.rb
index e3f4675c2..e974b8cd0 100644
--- a/stdlib/native.rb
+++ b/stdlib/native.rb
@@ -544,7 +544,11 @@ unless Hash.method_defined? :_initialize
           (value.constructor === undefined ||
             value.constructor === Object ||
             value instanceof Map)) {
-         $hash_put(hash, key, #{Hash.new(`value`)});
+          if (value.$$is_hash) {
+            $hash_put(hash, key, value);
+          } else {
+            $hash_put(hash, key, #{Hash.new(`value`)});
+          }
        } else if (value && value.$$is_array) {
          value = value.map(function(item) {
            if (item &&
```

I don't understand why `value instanceof Map` instead of `item instanceof Map` here:
https://github.com/opal/opal/blob/9d48ab627dc253ba26b1507caa768f772c361cbf/stdlib/native.rb#L553

**Comment by janbiedermann** [comment]
> I don't understand why `value instanceof Map` instead of `item instanceof Map` here:

Yes, me neither, thats probably a bug that i introduced. Fixed in #2671

**Comment by janbiedermann** [comment]
> +          if (value.$$is_hash) {
> +            $hash_put(hash, key, value);
> +          } else {
> +            $hash_put(hash, key, #{Hash.new(`value`)});
> +          }

First of all thanks for your report. Thinking about it i finally see the problem. 
The conversion that has been required previously, before Hash < Map, is no longer required with Hash < Map and the Map can be taken literally, which you correctly do in line 2 of your fix. This can be checked in the outer if, introducing a new branch there.
```
if (value && value instanceof Map) {
    $hash_put(hash, key, value);
} else if (value && value.constructor === undefined ||
             value.constructor === Object) {
```
Ill prepare a PR.

**Comment by lionelchauvin** [comment]
> The conversion that has been required previously, before Hash < Map, is no longer required with Hash < Map

Are you sure ? What happen when you have a Map containing a native javascript object ? With your proposal, I am not sure this native object will be converted to ruby object.

**Comment by janbiedermann** [comment]
Right, i updated #2672 to be recursive

**Comment by janbiedermann** [comment]
1.8 backport of #2672 available in #2680
